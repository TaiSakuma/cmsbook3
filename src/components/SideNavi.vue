<template>
  <v-card flat>
    <v-toolbar flat>
      <v-toolbar-title v-text="$route.params.chapter"></v-toolbar-title>
    </v-toolbar>
    <v-list shaped nav dense expand>
      <template v-for="page in pages">
        <template v-if="page.path">
          <v-list-item
            link
            router
            :key="page.name"
            :to="'/' + $route.params.chapter + '/' + page.path"
          >
            <v-list-item-action>
              <v-icon>mdi-book</v-icon>
            </v-list-item-action>
            <v-list-item-content>
              <v-list-item-title v-text="page.name"></v-list-item-title>
            </v-list-item-content>
          </v-list-item>
        </template>
        <template v-else-if="page.subcontents">
          <v-list-group prepend-icon="mdi-book-multiple" :key="page.name" >
            <template v-slot:activator>
              <v-list-item-title v-html="page.name"></v-list-item-title>
            </template>
            <template v-for="subpage in page.subcontents">
              <template v-if="subpage.path">
                <v-list-item
                  link
                  router
                  :key="subpage.name"
                  :to="'/' + $route.params.chapter + '/' + subpage.path"
                >
                  <v-list-item-action></v-list-item-action>
                  <v-list-item-content>
                    <v-list-item-title v-html="subpage.name"></v-list-item-title>
                  </v-list-item-content>
                  <v-list-item-icon>
                    <v-icon>mdi-book</v-icon>
                  </v-list-item-icon>
                </v-list-item>
              </template>
              <template v-else-if="subpage.subcontents">
                <v-list-group no-action sub-group :key="subpage.name">
                  <template v-slot:activator>
                    <v-list-item-content>
                      <v-list-item-title v-html="subpage.name"></v-list-item-title>
                    </v-list-item-content>
                  </template>
                  <v-list-item
                    link
                    router
                    v-for="subsubpage in subpage.subcontents"
                    :key="subsubpage.name"
                    :to="'/' + $route.params.chapter + '/' + subsubpage.path"
                  >
                    <v-list-item-title v-html="subsubpage.name"></v-list-item-title>
                    <v-list-item-icon>
                      <v-icon>mdi-book</v-icon>
                    </v-list-item-icon>
                  </v-list-item>
                </v-list-group>
              </template>
            </template>
          </v-list-group>
        </template>
      </template>
    </v-list>
  </v-card>
</template>

<script>
import axios from "axios";

export default {
  name: "SideNavi",
  data: () => ({
    pages: []
  }),
  methods: {
    updatePages() {
      let path = process.env.VUE_APP_CMSBOOK_URL;
      path = path + "/" + this.$route.params.chapter;
      path = path + "/.cmsbook3/sections.json";
      axios
        .get(path)
        .then(response => {
          this.pages = response.data.sections;
        })
        .catch(error => {
          this.pages = [];
        });
    }
  },
  created() {
    this.updatePages();
  },
  watch: {
    "$route.params.chapter": function() {
      this.updatePages();
    }
  }
};
</script>
