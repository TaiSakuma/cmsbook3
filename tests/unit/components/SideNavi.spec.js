import Vue from "vue";
import VueRouter from "vue-router";
import Vuex from "vuex";
import Vuetify from "vuetify";
import { mount, shallowMount, createLocalVue } from "@vue/test-utils";

import moxios from "moxios";

import SideNavi from "@/components/SideNavi.vue";

Vue.use(Vuetify);
// Vue.use(VueRouter);

describe("SideNavi.vue", () => {
  const ENV_ORG = process.env;

  let localVue;
  let router;
  let actions;
  let store;
  let wrapper;

  beforeEach(() => {
    process.env.VUE_APP_CMSBOOK_URL = "http://localhost/cmsbook";
    moxios.install();
    localVue = createLocalVue();
    localVue.use(Vuex);
    router = new VueRouter();

    actions = {};
    store = new Vuex.Store({
      actions,
      getters: {
        chapterMap: () => {
          return { "/chapter-A": { name: "Chapter A", path: "/chapter-A" } };
        },
      },
    });

    wrapper = shallowMount(SideNavi, {
      localVue,
      router,
      store,
      mocks: {
        $route: {
          path: "/chapter-A/section-b/page-3.md",
          params: {
            chapter: "chapter-A",
            section: "section-b",
            page: "page-3.md",
          },
        },
      },
    });
  });

  afterEach(() => {
    moxios.uninstall();
    process.env = ENV_ORG;
  });

  const response = {
    sections: [
      { name: "Inbox", path: "index/inbox.md" },
      {
        name: "A",
        subcontents: [
          { name: "a1", path: "a1" },
          { name: "a2", path: "a2" },
        ],
      },
      {
        name: "B",
        subcontents: [
          {
            name: "b1",
            subcontents: [
              { name: "b11", path: "b11" },
              { name: "b12", path: "b12" },
            ],
          },
          { name: "b2", path: "b2" },
        ],
      },
    ],
  };

  it("snapshot", (done) => {
    moxios.wait(async () => {
      let request = moxios.requests.mostRecent();
      await request.respondWith({
        status: 200,
        response: response,
      });
      expect(wrapper.html()).toMatchSnapshot();
      done();
    });
  });

  it("axios request ", (done) => {
    moxios.wait(() => {
      let request = moxios.requests.mostRecent();
      expect(request.config.url).toBe(
        "http://localhost/cmsbook/chapter-A/.cmsbook3/sections.json"
      );
      expect(request.config.method).toBe("get");
      expect(request.config.data).toBeUndefined();
      done();
    });
  });
});
