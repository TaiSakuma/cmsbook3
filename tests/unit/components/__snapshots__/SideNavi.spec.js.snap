// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideNavi.vue snapshot 1`] = `
<v-card-stub loaderheight="4" tag="div" flat="true">
  <v-toolbar-stub tag="header" extensionheight="48" flat="true" src="">
    <v-toolbar-title-stub>Chapter A</v-toolbar-title-stub>
  </v-toolbar-stub>
  <v-list-stub shaped="true" tag="div" dense="true" expand="true" nav="true">
    <v-list-item-stub activeclass="" link="true" to="/chapter-A/index/inbox.md" tag="div" router="">
      <v-list-item-action-stub>
        <v-icon-stub>mdi-book</v-icon-stub>
      </v-list-item-action-stub>
      <v-list-item-content-stub>
        <v-list-item-title-stub>Inbox</v-list-item-title-stub>
      </v-list-item-content-stub>
    </v-list-item-stub>
    <v-list-group-stub color="primary" activeclass="" appendicon="$expand" prependicon="mdi-book-multiple" ripple="true">
      <v-list-item-stub activeclass="" link="true" to="/chapter-A/a1" tag="div" router="">
        <v-list-item-action-stub></v-list-item-action-stub>
        <v-list-item-content-stub>
          <v-list-item-title-stub>a1</v-list-item-title-stub>
        </v-list-item-content-stub>
        <v-list-item-icon-stub>
          <v-icon-stub>mdi-book</v-icon-stub>
        </v-list-item-icon-stub>
      </v-list-item-stub>
      <v-list-item-stub activeclass="" link="true" to="/chapter-A/a2" tag="div" router="">
        <v-list-item-action-stub></v-list-item-action-stub>
        <v-list-item-content-stub>
          <v-list-item-title-stub>a2</v-list-item-title-stub>
        </v-list-item-content-stub>
        <v-list-item-icon-stub>
          <v-icon-stub>mdi-book</v-icon-stub>
        </v-list-item-icon-stub>
      </v-list-item-stub>
    </v-list-group-stub>
    <v-list-group-stub color="primary" activeclass="" appendicon="$expand" prependicon="mdi-book-multiple" ripple="true">
      <v-list-group-stub color="primary" activeclass="" appendicon="$expand" noaction="true" ripple="true" subgroup="true">
        <v-list-item-stub activeclass="" link="true" to="/chapter-A/b11" tag="div" router="">
          <v-list-item-title-stub>b11</v-list-item-title-stub>
          <v-list-item-icon-stub>
            <v-icon-stub>mdi-book</v-icon-stub>
          </v-list-item-icon-stub>
        </v-list-item-stub>
        <v-list-item-stub activeclass="" link="true" to="/chapter-A/b12" tag="div" router="">
          <v-list-item-title-stub>b12</v-list-item-title-stub>
          <v-list-item-icon-stub>
            <v-icon-stub>mdi-book</v-icon-stub>
          </v-list-item-icon-stub>
        </v-list-item-stub>
      </v-list-group-stub>
      <v-list-item-stub activeclass="" link="true" to="/chapter-A/b2" tag="div" router="">
        <v-list-item-action-stub></v-list-item-action-stub>
        <v-list-item-content-stub>
          <v-list-item-title-stub>b2</v-list-item-title-stub>
        </v-list-item-content-stub>
        <v-list-item-icon-stub>
          <v-icon-stub>mdi-book</v-icon-stub>
        </v-list-item-icon-stub>
      </v-list-item-stub>
    </v-list-group-stub>
  </v-list-stub>
</v-card-stub>
`;
